Cython/Utility/CppConvert.pyx

```diff
@@ -39,11 +39,16 @@ cdef inline object {{cname.replace("PyObject", py_type, 1)}}(const string& s):
 
 cdef extern from *:
     cdef cppclass vector "std::vector" [T]:
-        void push_back(T&)
+        ctypedef size_t size_type
+
+        void reserve(size_type) except +
+        void push_back(T&) except +
 
 @cname("{{cname}}")
 cdef vector[X] {{cname}}(object o) except *:
     cdef vector[X] v
+    if hasattr(o, '__len__'):
+        v.reserve(len(o))
```

Comment:
PEP 424, maybe?
https://www.python.org/dev/peps/pep-0424/

You could also check for the `sq_length` slot on the type directly, which is much faster than going through the generic `hasattr()`, especially for the most common types: containers (yes) and generators (no). `ObjectHandling.c` has examples where we use `sq_length()` directly.

Seems worth a separate PR.

Reply:
I will look into it.